gh-64192: Make `imap()`/`imap_unordered()` in `multiprocessing.pool` actually lazy

obaltian opened the pull request:
## Context recap (#64192)

Let's consider that we have an input `iterable` and `N = len(iterable)`.

Current `multiprocessing.Pool.imap` and `multiprocessing.Pool.imap_unordered` are *O(N)* in space (unecessarily expensive on large iterables, completely impossible to use on infinite iterables):
The call `results: Iterator = pool.imap(func, iterable)` iterates over all the elements of the iterable, submitting _N_ tasks to the `pool` (results are collected into a list of size _N_). Following calls to `next(results)` take the oldest result from the list (FIFO) (waiting for it if not available yet) and return it.

## Proposal: add an optional buffersize param

With this proposal, the call `results: Iterator = pool.imap(func, iterable, buffersize=b)` will iterate only over the first _b_ elements of `iterable` (acquiring `buffersize` semaphore while iterating), submitting _b_ tasks to worker threads and then will return the results iterator.

Calls to `next(results)` will release `buffersize` semaphore (allowing `task_handler` thread to get the next input element from `iterable` to submit a new task to a worker thread) and then return the result. 

`buffersize` semaphores from iterators not exhausted yet are also being released on pool termination to avoid deadlocks.

Benefits:

- The space complexity becomes _O(b)_
- When using a `buffersize` the client code takes back the control over the speed of iteration over the input `iterable`: after an initial spike of _b_ calls to `func` to fill the buffer, the iteration over input `iterable` will follow the rate of the iteration over the `results` (controlled by the client), which is critical when `func` involves talking to services that you don't want to overload.

## Feature history
`buffersize` support has been recently merged into `concurrent.futures.Executor.map` implementation (#125663) and many code/test/doc parts are based on ones from there to ensure consistency between modules. 
I want to thank authors of that PR for the references.

## Links:
* [Docs preview](https://cpython-previews--136871.org.readthedocs.build/en/136871/)
<!-- gh-issue-number: gh-64192 -->
* Issue: gh-64192
<!-- /gh-issue-number -->

Commit:
draft: impl lazy input consumption in mp.Pool.imap(_unordered)

Commit:
Use semaphore to synchronize threads

Using `threading.Semaphore` makes it easier to cap the number of
concurrently ran tasks. It also makes it possible to remove busy
wait in child thread by waiting for semaphore.

Also I've updated code to use the backpressure pattern - the new
tasks are scheduled as soon as the user consumes the old ones.

picnixz commented:
Please open an issue first for this one (I don't know if it's something that is discussed in the sprint).

picnixz commented:
Previous discussion for executor: https://github.com/python/cpython/issues/74028

obaltian commented:
@picnixz thanks. The issue already exists, I will add more detailed description here soon, just wanted to run all tests for now.

obaltian commented:
@picnixz here is a related issue - https://github.com/issues/recent?issue=python%7Ccpython%7C64192, I've left a summary at the end of that thread. Thank you.

picnixz commented:
Thanks! I'm actually adding the GH issue inthe title so that we can backref it easily.

Commit:
Update buffersize behavior to match concurrent.futures.Executor behavior

This new behavior allow smaller real concurrency number than
number of running processes. Previously, it was not allowed
since we implicitly incremented buffersize by `self._processes`.

Commit:
Release all `buffersize_lock` obj from the parent thread when terminate

Commit:
Add 2 basic `ThreadPool.imap()` tests w/ and w/o buffersize

Commit:
Fix accidental swap in imports

Commit:
clear Pool._taskqueue_buffersize_semaphores safely

Commit:
Slightly optimize Pool._taskqueue_buffersize_semaphores terminate

Commit:
Rename `Pool.imap()` buffersize-related tests

Commit:
Fix typo in `IMapIterator.__init__()`

Commit:
Add tests for buffersize combinations with other kwargs

Commit:
Remove if-branch in `_terminate_pool`

Commit:
Add more edge-case tests for `imap` and `imap_unodered`

These tests mostly come from a similar PR adding `buffersize` param
to `concurrent.futures.Executor.map` -
https://github.com/python/cpython/pull/125663/files

Commit:
Split inf iterable test for `imap` and `imap_unordered`

Commit:
Add doc for `buffersize` argument of `imap` and `imap_unordered`

obaltian commented:
Hello @ambv - we have talked regarding this PR during EuroPython sprints a week ago and I wanted to let you know that it's ready since (as you said you would review it).

Hello @gpshead - I'm also glad to say that it's ready to be reviewed, but we've already been discussing the idea with Łukasz so you can agree how do we proceed further.

Thanks in advance.

Commit:
add *versionadded* for `imap_unordered`

Commit:
Remove ambiguity in `buffersize` description.

Previously it was a bit ambigious to say
"pass this to fully utilize pool's capacity" as the fastest
way would still be not passing `buffersize` at all.
Now this section clearly says "... when using this feature".

obaltian commented on Lib/multiprocessing/pool.py:
```diff
@@ -382,72 +389,99 @@ def starmap_async(self, func, iterable, chunksize=None, callback=None,
             return (item for chunk in result for item in chunk)
 
-    def imap_unordered(self, func, iterable, chunksize=1):
+    def imap_unordered(self, func, iterable, chunksize=1, buffersize=None):
```
They are almost the same, indeed. Sure, I will make implementation shared, thanks.

picnixz replied:
I actually don't know if it's something that was thought to be duplicated. But it appears that it is. If @gpshead thinks it's better to duplicate the code for future improvements, then we can keep it as is (because if in the future only one of the two functions has issues or a different behavior, it's better to have two separate functions)

obaltian replied:
Another argument to reuse code is the fact that `_map_async` is also shared here.

obaltian replied:
I have added the shared method in a separate commit, can revert it if it's unwanted.

obaltian commented on Misc/NEWS.d/next/Library/2025-07-28-11-00-49.gh-issue-64192.7htLtg.rst:
```diff
@@ -0,0 +1,9 @@
+Add the optional ``buffersize`` parameter to
```
Sure, I'll add it. Just thought it's being written by the maintainers.

Commit:
Set *versionadded* as next in docs

Commit:
Add whatsnew entry

Commit:
Fix aggreed comments on code formatting/minor refactoring

Commit:
Remove `imap` and `imap_unordered` body code duplication

obaltian commented:
Hi @picnixz, letting you know I've fixed issues you mentioned, thanks.

obaltian commented:
Hello @gpshead, may I kindly ask if you plan this PR to be included in 3.15? The PR is ready and 2 threads are were left open to hear your opinion. 

Just wanted to clarify the process because we've discussed this feature with @ambv during sprints few weeks ago and he's supported it, saying he would review it in few weeks after the sprints, but he couldn't reply yet. 

Thank you.

obaltian committed:
Merge branch 'main' into feature/add-buffersize-to-multiprocessing

obaltian committed:
Merge branch 'main' into feature/add-buffersize-to-multiprocessing

obaltian committed:
Merge remote-tracking branch 'origin/main' into feature/add-buffersize-to-multiprocessing

Note: I also had to add 1 '=' character to `Doc/whatsnew/3.16.rst`
because check-merge-conflict pre-commit hook was detecting legit
markdown syntax as a merge conflict string.

obaltian committed:
Remove extra  import

encukou reviewed (commented):
Thank you for fixing this!

I think the new argument needs to be keyword-only -- passing it positionally would be quite confusing.

I'm afraid that as another reviewer, I'll add a few more smaller issues. Those aren't blocking, though.

obaltian committed:
Update Lib/multiprocessing/pool.py

Co-authored-by: Petr Viktorin <encukou@gmail.com>

obaltian committed:
Update Doc/library/multiprocessing.rst

Co-authored-by: Petr Viktorin <encukou@gmail.com>

obaltian committed:
Apply suggestion from @encukou

Co-authored-by: Petr Viktorin <encukou@gmail.com>

Commit:
Make `buffersize` argument kwarg-only

Commit:
Use set for `._taskqueue_buffersize_semaphores`

Commit:
Make `buffersize` arg check inline

Commit:
Simplify `.imap()` and `.imap_undered()` tests

Commit:
Fix iterator instantiation in in `.imap()`

Commit:
Mark kw-only arg in `imap`/`imap_unordered` in doc

obaltian committed:
Merge branch 'main' into feature/add-buffersize-to-multiprocessing

encukou reviewed (approved):
This looks good to me!
I'm not an expert on this module, so I'll give others a chance to chime in.
I'll merge next week if there are no objections.

encukou commented:
Sorry for the delay, I'm still catching up after EuroPython :)
